### ISSUE TITLE
TypeScript QueryHelpers return any instead of model type

### ISSUE THREAD

**Issue by florianwalther-private** [Automattic/mongoose#12342]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.5.3

### Node.js version

16.3.2

### MongoDB server version

5.0.11

### Description

I implemented a query helper function following the [tutorial documentation in the documentation](https://mongoosejs.com/docs/typescript/query-helpers.html).

As soon as I add this method, the return value of that query turns into `any`. It should return the proper model type instead.

### Steps to Reproduce

Implement a query helper as outlined in the documentation: https://mongoosejs.com/docs/typescript/query-helpers.html

### Expected Behavior

The return type after calling the query helper function should be the appropriate model type.

**Comment by hasezoey** [comment]
i think the mentioned documentation is at least slightly out-of-date, because a proper type for query-helpers should be `QueryWithHelpers` instead of just `Query`?

**Comment by florianwalther-private** [comment]
I'm not sure what the correct return type is because I find it a bit confusing, but it should certainly not be any. After calling `exec()`/`await`, I want the model back like it does without the query helper chained to it.

**Comment by hasezoey** [comment]
re https://github.com/Automattic/mongoose/issues/12347#issuecomment-1230463627

> but the referenced (my) issue is not about the documentation

from what i can tell, this issue is about the documentation because "you followed the documentation, but its not working" and after having taken a look at the documentation and the current code in mongoose's actual types, i came to the conclusion that the documentation is probably outdated

---

reason for the documentation being outdated:
since the documentation has been written `QueryWithHelpers` has been added, which should make the types easier & cleaner, see

https://github.com/Automattic/mongoose/blob/0af4cb815ec36dd50d18755f5238506ef9297a5a/types/query.d.ts#L24

by taking care of the `Query<...> & Type`

also the documentation lists the first parameter as a static `any` to `Query<any, ...>`, which is the result type, and so ending up in the wrong type

TL;DR: change your type to `QueryWithHelpers` for less code, and also change the first parameter from `any` to the actual return type

currently the documentation is basically guiding to be like (like in issue #12347)
https://github.com/Automattic/mongoose/blob/6cb1284821405c71a4d75e9857ef45d95c367972/types/models.d.ts#L405

here it should actually be `ResultDoc` or `{}` and definitely not `any`, see 

https://github.com/Automattic/mongoose/blob/0af4cb815ec36dd50d18755f5238506ef9297a5a/types/query.d.ts#L172

https://github.com/Automattic/mongoose/blob/0af4cb815ec36dd50d18755f5238506ef9297a5a/types/query.d.ts#L182-L184

TL;DR: if you are following the documentation, then you will be guaranteed to get `any` as a result after doing `query.exec()` (or `query.then` or implicitly like `await query`)

**Comment by florianwalther-private** [comment]
Thank you for the explanation. I didn't know that only the docs were wrong.

But I'm unsure what to pass instead of `any`. If I only pass the interface type I use for my model, it doesn't work. Instead, I copy pasted the return type I get from normal queries, which works, but is very long and ugly:

```
(mongoose.Document<unknown, any, User> & User & Required<{
    _id: mongoose.Types.ObjectId;
}>) | null
```

Can you give me any tips here?

Also, if I understand it correctly, QueryWithHelpers lets me remove the `TQueryHelpers` type argument. Is there any other difference? I noticed that my code works the same with and without this argument even if I use the normal `Query` type.

**Comment by hasezoey** [comment]
> Also, if I understand it correctly, QueryWithHelpers lets me remove the TQueryHelpers type argument. Is there any other difference? I noticed that my code works the same with and without this argument even if I use the normal Query type.

i basically suggested this because mongoose uses that internally over just `Query`, but i dont see other differences

> Instead, I copy pasted the return type I get from normal queries, which works, but is very long and ugly:

the expected return-type (if the query returns a single) should be a document (or document array), which should be done with `HydratedDocument<YourInterfaceType>`, also if you have this once you can re-use that argument for both the first and second generic

like (updated documentation):
```ts
// note the "T" generic is just to store the same result for both the first and second generic for "QueryWithHelpers"
type ProjectModelQuery<T = HydratedDocument<Project>> = QueryWithHelpers<T | null, T, ProjectQueryHelpers>;
// or
type ProjectModelQuery = QueryWithHelpers<HydratedDocument<Project> | null, HydratedDocument<Project>, ProjectQueryHelpers>;
```

**Comment by florianwalther-private** [comment]
@hasezoey Thank you so much! I think I tried `HydratedDocument` earlier but it didn't work because I didn't add the `| null`. Now it works!

**Comment by vkarpov15** [comment]
Fixed by #12507
